### ISSUE TITLE
Not valid Notification during rollback mode

### ISSUE THREAD

**Issue by proximax-core-dev** [nemtech/catapult-server#32]
If Notification's derived classes have a not primitive types like vector or unordered_map
(Or other types, which allocate memory inside on creation and free it on destruction),
it causes a not valid notifications after copy them by this way:

	// store a copy of the notification buffer
	const auto* pData = reinterpret_cast<const uint8_t*>(&notification);
	m_notificationBuffers.emplace_back(pData, pData + notification.Size);

To resolve it, we created a new class Cloneable, and inherited all Notifications from this class. Now we copy object by this way:

	// store a copy of the notification buffer
	m_notificationBuffers.emplace_back(notification.clone());

PR with our fix:
https://github.com/proximax-storage/cpp-xpx-catapult/pull/64

**Comment by gimer** [comment]
This is by-design.
Observer-channel `Notification`s aren't supposed to be using heap-allocated objects.
They should be using either value-type variables or references to objects owned _somewhere_.

Inside catapult the **only** notificiation that violate this rule is AddressInteractionNotification, but this is Validator-channel `Notification`. (there's appropriate \note there as well).

**Comment by gimer** [comment]
btw, link to pull request isn't accessible

**Comment by proximax-core-dev** [comment]
We understand that it is by-design. We added a new transactions and one of them required a notification which contains pointers to several arrays.
We noticed that AddressInteractionNotification already contains unordered_set and decided to add a vector to our Notification too. To avoid the bug with wrong copy we created a Cloneable class and inherited all Notifications from this class.

This solution allows to use STL containers inside of notifications, which can simplify code in some cases.

We agree that it is not required for current version of catapult, but the change is not hard and adds a new abilities.
